web/frontend/developing-ui.md

```diff
@@ -0,0 +1,183 @@
+.myComponent {
+  background-color: yellow;
+}
+
+p {
+  font-size: 16px;
+  color: blue;
+}
+
+
+// MyComponent.jsx
+import styles from './component.css'
+
+const MyComponent = () => <div className="myComponent"><p>This is a component!</p></div>
+```
+
+... it doesn’t matter whether or not that component is currently visible in the application -- that `p` style will apply to every `<p>` tag in your app on every page.
+
+That isn't to say you should _never_ write global CSS -- it is extremely useful when defining the default styles for the app, and the more CSS that can be shared and applied globally, the smaller your bundle size might be. However, if you are trying to write styles for a specific component or page, it is usually safest to use **CSS modules**.
+
+```
+// component.module.css
+.myComponent {
+  background-color: yellow;
+}
+
+p {
+  font-size: 16px;
+  color: blue;
+}
+
+// MyComponent.jsx
+import styles from './component.module.css'
+
+const MyComponent = () => <div className={styles.myComponent}><p>This is a component!</p></div>
+```
+
+The above code snippet looks very similar, but with one key difference: the file name ends in `.module.css`. That tells the F/E compiler to _hash_ any class names contained in that file. That means in the effective code, `myComponent` will end up looking something like `component_myComponent_1a2b3c`. Additionally, the `.module.css` file exports the hashed values as a JS object, so they can be referenced in JSX component code (see how `styles.myComponent` is used in the above snippet):
+
+```
+{
+  myComponent: 'component_myComponent_1a2b3c'
+}
+```
+
+The compiled CSS & HTML will end up looking like this:
+
+```
+// CSS
+.component_myComponent_1a2b3c {
+  background-color: yellow;
+}
+
+// HTML
+<div class="component_myComponent_1a2b3c"><p>This is a component!</p></div>
+```
+
+However, you might notice the `p` style block in that file will still apply globally, since it is not nested under any CSS class. So the CSS module file should finally be updated to:
+
+```
+// component.module.css
+.myComponent {
+  background-color: yellow;
+}
+
+.myComponent p {
+  font-size: 16px;
+  color: blue;
+}
+```
+
+All of the above also applies to SASS/SCSS files (`.scss, .module.scss`).
+
+### References
+
+- [CSS modules](https://github.com/css-modules/css-modules)
+- [What are CSS modules and why do we need them?](https://css-tricks.com/css-modules-part-1-need/)
+
+## CSS: less is more
+
+Another reason I prefer to start with defining the HTML is because you will immediately get to see how far that gets you without writing any CSS at all. Most of the time, base styles such as typography and layout should be inherited by global CSS or top-level components. If additional styling on a page or component is needed, it's usually a good idea to verify with the designer whether it is a change that should occur globally or not, and if not then that's usually a signal that a new UI component is needed.
```

Comment:
Should this be from one person's perspective ("I prefer") or is this coming from the general Truss voice?

Reply:
whoops, more copy paste